File: 46_Permutations.go
func permute(nums []int) [][]int {
    result := [][]int {}
    dfs(&result, nums, 0, len(nums))    
    return result
}

func dfs (result *[][] int, nums [] int, begin int, size int) {
    
    if begin >= size {
        temp := make([] int, size)
        copy(temp, nums)
        *result = append(*result, temp)
        fmt.Println(*result)
        return
    }
    
    for i := begin; i < size; i++ {
        nums[begin], nums[i] = nums[i], nums[begin]
        fmt.Println("nums : ", nums, ", begin : ", begin, ", i : ", i)
        dfs(result, nums, begin + 1, size)
        nums[begin], nums[i] = nums[i], nums[begin]
    } 
}

File: 46_Permutations.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
    const result = [];
    
    const backtracking = (curr, rest) => {
        if(rest.length === 0) {
            result.push(curr);
            return;
        }
        
        for(let i = 0; i < rest.length; ++i) {
            let print1 = [...curr, rest[i]]
            let print2 = [...rest.slice(0, i), ...rest.slice(i + 1)]
            console.log(print1, print2)
            backtracking(print1, print2)
        }
    }

    backtracking([], nums);
    return result;
};